libfwupdplugin/fu-bluez-device.c

```diff
@@ -602,6 +814,106 @@ fu_bluez_device_notify_stop(FuBluezDevice *self, const gchar *uuid, GError **err
 	return TRUE;
 }
 
+static gboolean
+fu_bluez_device_method_acquire(FuBluezDevice *self,
+			       const gchar *method,
+			       const gchar *uuid,
+			       gint *fd,
+			       gint32 *mtu,
+			       GError **error)
+{
+	FuBluezDeviceUuidHelper *uuid_helper;
+	GVariant *opt_variant = NULL;
+	gint fd_id;
+	g_autoptr(GVariantBuilder) opt_builder = NULL;
+	g_autoptr(GVariant) val = NULL;
+	g_autoptr(GUnixFDList) out_fd_list = NULL;
+
+	uuid_helper = fu_bluez_device_get_uuid_helper(self, uuid, error);
+	if (uuid_helper == NULL)
+		return FALSE;
+	if (!fu_bluez_device_ensure_uuid_helper_proxy(uuid_helper, error))
+		return FALSE;
+
+	opt_builder = g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+	opt_variant = g_variant_new("a{sv}", opt_builder);
+
+	val = g_dbus_proxy_call_with_unix_fd_list_sync(uuid_helper->proxy,
+						       method,
+						       g_variant_new("(@a{sv})", opt_variant),
+						       G_DBUS_CALL_FLAGS_NONE,
+						       -1,
+						       NULL, // fd list
+						       &out_fd_list,
+						       NULL,
+						       error);
+	if (val == NULL) {
+		g_prefix_error(error, "failed to call %s: ", method);
+		return FALSE;
+	}
+
+	g_variant_get_child(val, 0, "h", &fd_id);
+	g_variant_get_child(val, 1, "q", mtu);
+
+	*fd = g_unix_fd_list_get(out_fd_list, fd_id, NULL);
+
+	return TRUE;
+}
+
+/**
+ * fu_bluez_device_notify_acquire:
+ * @self: a #FuBluezDevice
+ * @uuid: the UUID, e.g. `00cde35c-7062-11eb-9439-0242ac130002`
+ * @fd: (out) file descriptor for direct communication with the
+ *      device
+ * @mtu: (out) MTU of the channel
+ * @error: (nullable): optional return location for an error
+ *
+ * Acquire notifications for property changes in a UUID (AcquireNotify
+ * method).
+ * Closing file descriptor releases the notify.
+ *
+ * Returns: %TRUE if the method call completed successfully.
+ *
+ * Since: 2.0.0
+ **/
+gboolean
+fu_bluez_device_notify_acquire(FuBluezDevice *self,
+			       const gchar *uuid,
+			       gint *fd,
```

Comment:
rather than returning a "bare" FD, could we return a GInputStream?

Reply:
@hughsie maybe FuIOChannel?